File: PracticasProgramacion/Practica_23/python/matriz_autoreferenciada.py
"""
MatrizAutoreferenciada: contiene siguiente_ como referencia al mismo tipo.
Producto matricial con la matriz apuntada por siguiente_.
"""

from typing import Optional


class MatrizAutoreferenciada:
    def __init__(self, filas: int, columnas: int) -> None:
        self._filas = filas
        self._columnas = columnas
        self._datos = [[0.0] * columnas for _ in range(filas)]
        self._siguiente: Optional[MatrizAutoreferenciada] = None

    def set_siguiente(self, otra: "MatrizAutoreferenciada") -> None:
        self._siguiente = otra

    def get_siguiente(self) -> Optional["MatrizAutoreferenciada"]:
        return self._siguiente

    def llenar(self, valor: float) -> None:
        for i in range(self._filas):
            for j in range(self._columnas):
                self._datos[i][j] = valor

    def multiplicar_por_constante(self, k: float) -> "MatrizAutoreferenciada":
        r = MatrizAutoreferenciada(self._filas, self._columnas)
        for i in range(self._filas):
            for j in range(self._columnas):
                r._datos[i][j] = self._datos[i][j] * k
        return r

    def multiplicar_con_siguiente(self) -> "MatrizAutoreferenciada":
        if self._siguiente is None:
            raise ValueError("No hay matriz siguiente")
        B = self._siguiente
        if self._columnas != B._filas:
            raise ValueError("Dimensiones incompatibles")
        r = MatrizAutoreferenciada(self._filas, B._columnas)
        for i in range(self._filas):
            for j in range(B._columnas):
                s = 0.0
                for k in range(self._columnas):
                    s += self._datos[i][k] * B._datos[k][j]
                r._datos[i][j] = s
        return r

    def imprimir(self) -> None:
        for i in range(self._filas):
            print(" ".join(str(self._datos[i][j]) for j in range(self._columnas)))

File: PracticasProgramacion/Practica_23/cpp/MatrizAutoreferenciada.cpp
#include "MatrizAutoreferenciada.h"
#include <iostream>
#include <stdexcept>

MatrizAutoreferenciada::MatrizAutoreferenciada(int filas, int columnas)
    : filas_(filas), columnas_(columnas), siguiente_(nullptr) {
    datos_ = new double*[filas_];
    for (int i = 0; i < filas_; ++i)
        datos_[i] = new double[columnas_]();
}

MatrizAutoreferenciada::MatrizAutoreferenciada(const MatrizAutoreferenciada& otra)
    : filas_(otra.filas_), columnas_(otra.columnas_), siguiente_(nullptr) {
    datos_ = new double*[filas_];
    for (int i = 0; i < filas_; ++i) {
        datos_[i] = new double[columnas_];
        for (int j = 0; j < columnas_; ++j)
            datos_[i][j] = otra.datos_[i][j];
    }
}

MatrizAutoreferenciada& MatrizAutoreferenciada::operator=(const MatrizAutoreferenciada& otra) {
    if (this != &otra) {
        liberar();
        filas_ = otra.filas_;
        columnas_ = otra.columnas_;
        siguiente_ = nullptr;
        datos_ = new double*[filas_];
        for (int i = 0; i < filas_; ++i) {
            datos_[i] = new double[columnas_];
            for (int j = 0; j < columnas_; ++j)
                datos_[i][j] = otra.datos_[i][j];
        }
    }
    return *this;
}

MatrizAutoreferenciada::~MatrizAutoreferenciada() { liberar(); }

void MatrizAutoreferenciada::liberar() {
    if (datos_) {
        for (int i = 0; i < filas_; ++i)
            delete[] datos_[i];
        delete[] datos_;
        datos_ = nullptr;
    }
}

void MatrizAutoreferenciada::setSiguiente(MatrizAutoreferenciada* otra) {
    siguiente_ = otra;
}

MatrizAutoreferenciada* MatrizAutoreferenciada::getSiguiente() const {
    return siguiente_;
}

void MatrizAutoreferenciada::llenar(double valor) {
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < columnas_; ++j)
            datos_[i][j] = valor;
}

MatrizAutoreferenciada MatrizAutoreferenciada::multiplicarPorConstante(double k) const {
    MatrizAutoreferenciada R(filas_, columnas_);
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < columnas_; ++j)
            R.datos_[i][j] = datos_[i][j] * k;
    return R;
}

MatrizAutoreferenciada MatrizAutoreferenciada::multiplicarConSiguiente() const {
    if (!siguiente_)
        throw std::runtime_error("No hay matriz siguiente");
    const MatrizAutoreferenciada& B = *siguiente_;
    if (columnas_ != B.filas_)
        throw std::runtime_error("Dimensiones incompatibles");
    MatrizAutoreferenciada R(filas_, B.columnas_);
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < B.columnas_; ++j) {
            double s = 0;
            for (int k = 0; k < columnas_; ++k)
                s += datos_[i][k] * B.datos_[k][j];
            R.datos_[i][j] = s;
        }
    return R;
}

void MatrizAutoreferenciada::imprimir() const {
    for (int i = 0; i < filas_; ++i) {
        for (int j = 0; j < columnas_; ++j)
            std::cout << datos_[i][j] << " ";
        std::cout << "\n";
    }
}
